From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fiona Ebner <f.ebner@proxmox.com>
Date: Wed, 30 Sep 2026 15:43:54 +0200
Subject: [PATCH] PVE backup: vma reader: restrict allowed archive member names
 for restore

When restoring an archive, member names were not restricted. Disallow
member names that are not valid simple path components.

For configuration members, the name is treated as a component of the
target path, so the restriction is important.

For device members, the protection is hardening/future-proofing,
because the name is used as a non-initial substring of a component of
the target part.

Reported-by: Jakob Klocker <j.klocker@proxmox.com>
Signed-off-by: Fiona Ebner <f.ebner@proxmox.com>
Reviewed-by: Jakob Klocker <j.klocker@proxmox.com>
Tested-by: Jakob Klocker <j.klocker@proxmox.com>
---
 vma-reader.c | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/vma-reader.c b/vma-reader.c
index ab0fe6f3b7..98951e6f95 100644
--- a/vma-reader.c
+++ b/vma-reader.c
@@ -51,6 +51,21 @@ struct VmaReader {
     int clusters_read_per;
 };
 
+static bool vma_check_member_name(const char *member_name, Error **errp)
+{
+    if (strchr(member_name, '/')) {
+        error_setg(errp, "must not contain '/' - ('%s')", member_name);
+        return false;
+    } else if (strcmp(member_name, "..") == 0) {
+        error_setg(errp, "must not equal '..'");
+        return false;
+    } else if (strcmp(member_name, ".") == 0) {
+        error_setg(errp, "must not equal '.'");
+        return false;
+    }
+    return true;
+}
+
 static guint
 g_int32_hash(gconstpointer v)
 {
@@ -321,6 +336,11 @@ static int vma_reader_read_head(VmaReader *vmar, Error **errp)
         const char *devname =  get_header_str(vmar, devname_ptr);
 
         if (size && devname) {
+            if (!vma_check_member_name(devname, errp)) {
+                error_prepend(errp, "archive contains invalid device name: ");
+                return -1;
+            }
+
             vmar->devinfo[i].size = size;
             vmar->devinfo[i].devname = devname;
 
@@ -345,6 +365,11 @@ static int vma_reader_read_head(VmaReader *vmar, Error **errp)
             return -1;
         }
 
+        if (!vma_check_member_name(name, errp)) {
+            error_prepend(errp, "archive contains invalid config name: ");
+            return -1;
+        }
+
         VmaConfigData *cdata = g_new0(VmaConfigData, 1);
         cdata->name = name;
         cdata->data = blob->data;
